File: src/domain/display_prefix.zig
//! Derives a Local Display ID prefix from a repository basename.
//!
//! Algorithm (per `ARCHITECTURE.md` ("Repository Store Contracts")):
//! - Lowercase.
//! - Treat underscores as separators.
//! - Split on separators and punctuation (`-`, `_`, `.`, `/`, `:`, `#`,
//!   whitespace).
//! - Drop empty segments.
//! - If the joined form (segments joined with `-`) is at most 12 chars, use it.
//! - Else if the first two segments joined with `-` fit in 12 chars, use that.
//! - Else truncate the sanitized basename to 12 chars.
//! - If the result is empty or starts with a digit, prefix with `tk-`.
//!
//! Vowels are not stripped. Output is always lowercase ASCII.

const std = @import("std");
const Allocator = std.mem.Allocator;

/// Maximum length of the stored local Display ID prefix.
pub const max_prefix_len: usize = 12;

const separators = "-_./:# \t";

/// Returns a newly-allocated, lowercase prefix derived from `basename`.
/// Caller owns the returned slice.
pub fn derive(gpa: Allocator, basename: []const u8) ![]u8 {
    const lowered = try std.ascii.allocLowerString(gpa, basename);
    defer gpa.free(lowered);

    var segments: std.ArrayList([]u8) = .empty;
    defer {
        for (segments.items) |seg| gpa.free(seg);
        segments.deinit(gpa);
    }

    var it = std.mem.tokenizeAny(u8, lowered, separators);
    while (it.next()) |raw| {
        const seg = try filterAlnum(gpa, raw);
        if (seg.len == 0) {
            gpa.free(seg);
            continue;
        }
        try segments.append(gpa, seg);
    }

    const result = try chooseShape(gpa, segments.items);
    if (result.len == 0 or std.ascii.isDigit(result[0])) {
        defer gpa.free(result);
        return std.fmt.allocPrint(gpa, "tk-{s}", .{result});
    }
    return result;
}

fn chooseShape(gpa: Allocator, segments: []const []u8) ![]u8 {
    const joined_all = try std.mem.join(gpa, "-", segments);
    if (joined_all.len > 0 and joined_all.len <= max_prefix_len) return joined_all;
    defer gpa.free(joined_all);

    if (segments.len >= 2) {
        const joined_two = try std.mem.join(gpa, "-", segments[0..2]);
        if (joined_two.len <= max_prefix_len) return joined_two;
        gpa.free(joined_two);
    }

    const concat = try std.mem.concat(gpa, u8, segments);
    defer gpa.free(concat);
    const upper = @min(concat.len, max_prefix_len);
    return gpa.dupe(u8, concat[0..upper]);
}

fn filterAlnum(gpa: Allocator, segment: []const u8) ![]u8 {
    var out: std.ArrayList(u8) = .empty;
    errdefer out.deinit(gpa);
    for (segment) |b| {
        if (std.ascii.isAlphanumeric(b)) try out.append(gpa, b);
    }
    return out.toOwnedSlice(gpa);
}

test "derive: short single word" {
    const out = try derive(std.testing.allocator, "ticket");
    defer std.testing.allocator.free(out);
    try std.testing.expectEqualStrings("ticket", out);
}

test "derive: lowercases" {
    const out = try derive(std.testing.allocator, "Ticket");
    defer std.testing.allocator.free(out);
    try std.testing.expectEqualStrings("ticket", out);
}

test "derive: multi-word fits joined" {
    const out = try derive(std.testing.allocator, "my-cool-app");
    defer std.testing.allocator.free(out);
    try std.testing.expectEqualStrings("my-cool-app", out);
}

test "derive: long basename falls back to first two segments" {
    const out = try derive(std.testing.allocator, "src-cafe-extras-and-more");
    defer std.testing.allocator.free(out);
    try std.testing.expectEqualStrings("src-cafe", out);
}

test "derive: very long single word truncates to 12" {
    const out = try derive(std.testing.allocator, "abcdefghijklmnopqrstuvwxyz");
    defer std.testing.allocator.free(out);
    try std.testing.expectEqualStrings("abcdefghijkl", out);
}

test "derive: underscores treated as separators" {
    const out = try derive(std.testing.allocator, "my_cool_app");
    defer std.testing.allocator.free(out);
    try std.testing.expectEqualStrings("my-cool-app", out);
}

test "derive: punctuation stripped from segments" {
    const out = try derive(std.testing.allocator, "ab.cd!ef");
    defer std.testing.allocator.free(out);
    try std.testing.expectEqualStrings("ab-cdef", out);
}

test "derive: empty becomes tk-" {
    const out = try derive(std.testing.allocator, "");
    defer std.testing.allocator.free(out);
    try std.testing.expectEqualStrings("tk-", out);
}

test "derive: all-punctuation becomes tk-" {
    const out = try derive(std.testing.allocator, "---");
    defer std.testing.allocator.free(out);
    try std.testing.expectEqualStrings("tk-", out);
}

test "derive: digit-leading prefixed with tk-" {
    const out = try derive(std.testing.allocator, "42-things");
    defer std.testing.allocator.free(out);
    try std.testing.expectEqualStrings("tk-42-things", out);
}

test "derive: vowels preserved" {
    const out = try derive(std.testing.allocator, "iou");
    defer std.testing.allocator.free(out);
    try std.testing.expectEqualStrings("iou", out);
}

test "derive: long first two segments truncate full basename" {
    const out = try derive(std.testing.allocator, "alphabetone-betatwo-gamma");
    defer std.testing.allocator.free(out);
    try std.testing.expectEqualStrings("alphabetoneb", out);
}

File: crates/tk/src/store/display_prefix.rs
//! Derives a Local Display ID prefix from a repository basename.
//!
//! Algorithm (per `ARCHITECTURE.md`, "Repository Store Contracts"):
//!
//! - Lowercase.
//! - Treat underscores as separators (so `my_cool_app` → `my-cool-app`).
//! - Split on separators and punctuation (`-`, `_`, `.`, `/`, `:`, `#`,
//!   whitespace).
//! - Drop empty segments.
//! - If the joined form (segments joined with `-`) is at most 12 chars, use it.
//! - Else if the first two segments joined with `-` fit in 12 chars, use that.
//! - Else truncate the concatenated sanitized basename to 12 chars.
//! - If the result is empty or starts with a digit, prefix with `tk-`.
//!
//! Vowels are not stripped. Output is always lowercase ASCII.

/// Maximum length of the stored local Display ID prefix.
pub const MAX_PREFIX_LEN: usize = 12;

const SEPARATORS: &[char] = &['-', '_', '.', '/', ':', '#', ' ', '\t'];

/// Derive a Display ID prefix from `basename`. Always returns lowercase ASCII.
#[must_use]
pub fn derive(basename: &str) -> String {
    let lowered = basename.to_ascii_lowercase();
    let segments: Vec<String> = lowered
        .split(|c| SEPARATORS.contains(&c))
        .filter_map(|raw| {
            let filtered: String = raw.chars().filter(char::is_ascii_alphanumeric).collect();
            if filtered.is_empty() {
                None
            } else {
                Some(filtered)
            }
        })
        .collect();

    let result = choose_shape(&segments);
    if result.is_empty() || result.starts_with(|c: char| c.is_ascii_digit()) {
        return format!("tk-{result}");
    }
    result
}

fn choose_shape(segments: &[String]) -> String {
    let joined_all = segments.join("-");
    if !joined_all.is_empty() && joined_all.len() <= MAX_PREFIX_LEN {
        return joined_all;
    }
    if segments.len() >= 2 {
        let joined_two = segments[..2].join("-");
        if joined_two.len() <= MAX_PREFIX_LEN {
            return joined_two;
        }
    }
    let concat: String = segments.concat();
    let cap = concat.len().min(MAX_PREFIX_LEN);
    concat[..cap].to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn short_single_word() {
        assert_eq!(derive("ticket"), "ticket");
    }

    #[test]
    fn lowercases() {
        assert_eq!(derive("Ticket"), "ticket");
    }

    #[test]
    fn multi_word_fits_joined() {
        assert_eq!(derive("my-cool-app"), "my-cool-app");
    }

    #[test]
    fn long_basename_falls_back_to_first_two_segments() {
        assert_eq!(derive("src-cafe-extras-and-more"), "src-cafe");
    }

    #[test]
    fn very_long_single_word_truncates_to_12() {
        assert_eq!(derive("abcdefghijklmnopqrstuvwxyz"), "abcdefghijkl");
    }

    #[test]
    fn underscores_treated_as_separators() {
        assert_eq!(derive("my_cool_app"), "my-cool-app");
    }

    #[test]
    fn punctuation_stripped_from_segments() {
        assert_eq!(derive("ab.cd!ef"), "ab-cdef");
    }

    #[test]
    fn empty_becomes_tk_dash() {
        assert_eq!(derive(""), "tk-");
    }

    #[test]
    fn all_punctuation_becomes_tk_dash() {
        assert_eq!(derive("---"), "tk-");
    }

    #[test]
    fn digit_leading_prefixed_with_tk_dash() {
        assert_eq!(derive("42-things"), "tk-42-things");
    }

    #[test]
    fn vowels_preserved() {
        assert_eq!(derive("iou"), "iou");
    }

    #[test]
    fn long_first_two_segments_truncate_full_basename() {
        assert_eq!(derive("alphabetone-betatwo-gamma"), "alphabetoneb");
    }
}
